File: problems/leetcode/0500/560.cpp
// 560. Subarray Sum Equals K
// https://leetcode.com/problems/subarray-sum-equals-k/

#include <map>
#include <vector>

using namespace std;

class Solution {
 public:
  int subarraySum(vector<int>& nums, int k) {
    int count = 0, sum = 0;
    map<int, int> numberOfCountByPrefixSum;
    numberOfCountByPrefixSum.insert(make_pair(0, 1));

    for (auto& num : nums) {
      sum += num;

      if (numberOfCountByPrefixSum.find(sum - k) !=
          numberOfCountByPrefixSum.end()) {
        count += numberOfCountByPrefixSum[sum - k];
      }

      if (numberOfCountByPrefixSum.find(sum) !=
          numberOfCountByPrefixSum.end()) {
        numberOfCountByPrefixSum[sum] += 1;
      } else {
        numberOfCountByPrefixSum.insert(make_pair(sum, 1));
      }
    }

    return count;
  }
};

File: problems/leetcode/0500/560/560.py
# 560. Subarray Sum Equals K
# https://leetcode.com/problems/subarray-sum-equals-k/description/

from collections import defaultdict
from typing import *

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefixSums = defaultdict(int)
        prefixSums[0] += 1
        totalSum, totalCount = 0, 0

        for num in nums:
            totalSum += num
            totalCount += prefixSums[totalSum - k]
            prefixSums[totalSum] += 1

        return totalCount
